src/AudioPipe.cpp

```diff
@@ -254,17 +232,85 @@ AudioBuffer::shared AudioPipe::RecBuffer(DWORD frameSize)
 			//Exit
 			Log("AudioPipe: RecBuffer cancelled.\n");
 			//End
-			goto end;
+			pthread_mutex_unlock(&mutex);
+			return {};
 		}
 	}
-	//Get samples from queue
-	len = rateBlockBuffer.pop(buffer, totalSize) / numChannels;
-end:
-	//Unlock
+	// since number of samples in decoded audio buffer and encoder frame size may differ,
+	// rateBlockBuffer is used to store the leftover. So always read number of samples
+	// from rateBlockBuffer first, and if more samples are needed, then read from queue.
+	AudioBuffer::const_shared frontAudioBuffer;
+	auto bufferedSamples = rateBlockBuffer.length() / numChannels;
+	// readSamples: num samples per channel read from rateBlockBuffer
+	auto readSamples = std::min(bufferedSamples, frameSize);
+	// read whatever in rateBlockBuffer first
+	rateBlockBuffer.pop(buffer, readSamples*numChannels);
+	// update write offset since readSamples * numChannels already written into the buffer
+	size_t offset = readSamples * numChannels;
+	// needSamples: num samples per channel needed from queue after consuming rateBlockBuffer
+	int needSamples = frameSize - readSamples;
```

Comment:
reuse `requiredSamples`
```suggestion
	requiredSamples -= readSamples;
```